### PR TITLE
File upload support

### PR DESCRIPTION
**Pull request by ivelten**
With this PR, I am adding an experimental tool package for working with AspNetCore using this [multipart request spec](https://github.com/jaydenseric/graphql-multipart-request-spec) to enable a file upload feature using GraphQL.

The idea behind the component is to provide a MultipartRequest object that will be parsed from a MultipartReader. This parsing operation will automatically fill the file into the variables object, enabling user code to fetch the stream from the resolver function of a mutation field, for example.

- [x] Create an input object for file upload
- [x] Create a MultipartRequest object
- [x] Create a routine to read multipart requests and parse them to the MultiparseRequest object
- [x] Automated tests
- [x] Update sample server
- [x] Create nuget package
- [x] Create documentation

### CODE REVIEW

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (Horusiath) [inline src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: Use `Seq.choose` instead:

```fsharp
map
|> Seq.choose (fun x -> x.Value |> Seq.tryHead |> Option.map (fun v -> (k, v))
|> Map.ofSeq
```

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (Horusiath) [inline src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: You can use `KeyValue` acive pattern: `Seq.map (|KeyValue|)`.

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (Horusiath) [inline src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: Keep in mind that sprintf is an expensive operation.

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (Horusiath) [inline src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: `Async.RunSynchronously`? (┛◉Д◉)┛彡┻━┻

**Review by Horusiath — CHANGES_REQUESTED** [review]
Few minor stylish issues, and few major ones (mostly around the area of blocking run of asyncs and sprintf). 

Also I'm worried about relying on JSON.NET more and more, but as this is a companion library which already has ASP.NET Core dependencies, I think we're eventually ok with that.

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (ivelten) [reply src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: I switched that for `String.Format`, as it seems better by this [comparison](http://www.fssnip.net/mQ).

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (ivelten) [reply src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: OK, true, that was bad. Solved.

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (ivelten) [reply src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: Done.

File: src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs
Comment (ivelten) [reply src/FSharp.Data.GraphQL.Server.Middlewares.AspNetCore/MultipartRequest.fs]: Done.

**Comment by ivelten** [comment]
@Horusiath I did all the changes you asked for. About JSON.NET dependency, yeah, as this component is designed for ASP.NET Core and Microsoft already adopted this library, I don't think this is an issue in this case.

**Comment by Rajivhost** [comment]
Apparently done, I'm waiting this to be merged :)

**Comment by xperiandri** [comment]
@ivelten, does this functionality work?
Could you rebase it onto the latest dev?

**Comment by ivelten** [comment]
@xperiandri it was working when I finished the PR. If everyone decides to restart this PR, I can try rebasing and/or adjusting minor issues on it next weekend.

**Comment by xperiandri** [comment]
@aggieben, @johnberzy-bazinga, any objections?

**Comment by aggieben** [comment]
I don't object to the PR, but I'd think you'd want to rebase before merging

**Comment by johnberzy-bazinga** [comment]
I'd hold off on rebasing this. I'm working on a middleware that uses `System.Text.Json` that also supports File uploads.  The goal is to remove the boilerplate by supporting something like:

`app.UseGraphQL(...)`

It's close to being done, but on problem I ran into with `System.Text.Json` is it doesn't have anything analogues to `JObject.FromObject` so it's trickier to support the replacements by jsonpath that are needed for multipart file uploads.

**Comment by xperiandri** [comment]
I suppose that according to the sample in this repo it is already done, right?

**Comment by aggieben** [comment]
Don't we want the server-side implementation in this PR?

**Comment by xperiandri** [comment]
It was there as a sample but I broke it when I implemented getting variables from JsonDocument

**Pull request closed without merging**
